File: src/Thermostat.js
'use strict';

function Thermostat() {
  this._MINIMUM_TEMPERATURE = 10;
  this._MAX_LIMIT_PSM_ON = 25;
  this._MAX_LIMIT_PSM_OFF = 32;
  this._DEFAULT_TEMPERATURE = 20;
  this._TEMPERATURE_CHANGE = 1;
  this._MEDIUM_ENERGY_USAGE_LIMIT = 18;
  this._temperature = this._DEFAULT_TEMPERATURE;
  this._powerSavingMode = true;
}

Thermostat.prototype = {
  getCurrentTemperature: function() {
    return this._temperature;
  },
  up: function() {
    if (this._isMaximumTemperature()) {
      return;
    }
    this._temperature += this._TEMPERATURE_CHANGE;
  },
  down: function() {
    if (this._isMinimumTemperature()) {
      return;
    }
    this._temperature -= this._TEMPERATURE_CHANGE;
  },
  isPowerSavingModeOn: function() {
    return this._powerSavingMode;
  },
  switchPowerSavingModeOff: function() {
    this._powerSavingMode = false;
  },
  switchPowerSavingModeOn: function() {
    this._powerSavingMode = true;
    if (this._temperature > this._MAX_LIMIT_PSM_ON) {
      this._temperature = this._MAX_LIMIT_PSM_ON;
    }
  },
  resetTemperature: function() {
    this._temperature = this._DEFAULT_TEMPERATURE;
  },
  energyUsage: function() {
    if (this._temperature < this._MEDIUM_ENERGY_USAGE_LIMIT) {
      return 'low-usage';
    }
    if (this._temperature < this._MAX_LIMIT_PSM_ON) {
      return 'medium-usage';
    }
    return 'high-usage';
  },
  _isMinimumTemperature: function() {
    return this._temperature == this._MINIMUM_TEMPERATURE;
  },
  _isMaximumTemperature: function() {
    if (this.isPowerSavingModeOn()) {
      return this._temperature == this._MAX_LIMIT_PSM_ON;
    }
    return this._temperature == this._MAX_LIMIT_PSM_OFF
  }
};

File: lib/thermostat.rb
class Thermostat
  attr_reader :temperature

  DEFAULT_TEMPERATURE = 20

  def initialize
    @temperature = DEFAULT_TEMPERATURE
    @power_save_mode = true
  end

  def up
    @temperature += 1 unless maximum_reached?
  end

  def down
    @temperature -= 1 unless minimum_reached?
  end

  def reset
    @temperature = DEFAULT_TEMPERATURE
  end

  def self.instance
    @thermostat ||= self.new
  end

  def in_power_save_mode?
    @power_save_mode
  end

  def power_save_off
    @power_save_mode = false
  end

  def power_save_on
    @power_save_mode = true
    @temperature = 25 if maximum_reached?
  end

  def energy_usage
    if temperature < 18
      return :low
    elsif temperature < 25
      return :medium
    end
    return :high
  end

  private

  def maximum_reached?
    if in_power_save_mode?
      temperature >= 25
    else
      temperature == 32
    end
  end

  def minimum_reached?
    temperature == 10
  end

end
